File: lintcode/python/0093_balanced_binary_tree.py
#!/usr/bin/python -t

# binary tree
# divid and conquer way
# time O(n)
# space O(1)


class Solution:
    """
    @param root: The root of binary tree.
    @return: True if this Binary tree is Balanced, or false.
    """
    def is_balanced(self, root: TreeNode) -> bool:
        # write your code here
        if not root:
            return True

        balanced, _ = self.helper(root)

        return balanced

    def helper(self, node):
        if not node:
            return True, 0

        l_b, l_d = self.helper(node.left)
        r_b, r_d = self.helper(node.right)

        balanced = l_b and r_b and abs(l_d -r_d) < 2
        depth = max(l_d, r_d) + 1

        return balanced, depth

# dfs, divid and conquer solution

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: The root of binary tree.
    @return: True if this Binary tree is Balanced, or false.
    """
    def isBalanced(self, root):
        # write your code here
        if root == None:
            return True
            
        balanced, _ = self.dfs(root)
        
        return balanced
        
    def dfs(self, node):
        if node == None:
            return True, 0
            
        l_b, l_d = self.dfs(node.left)
        if not l_b:
            return False, 0
            
        r_b, r_d = self.dfs(node.right)
        if not r_b:
            return False, 0
            
        return abs(l_d-r_d) <= 1, max(l_d, r_d)+1

File: lintcode/golang/0093_balanced_binary_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: The root of binary tree.
 * @return: True if this Binary tree is Balanced, or false.
 */
func isBalanced (root *TreeNode) bool {
    // write your code here
    if root == nil {
        return true
    }
    
    var helper func(*TreeNode) (bool, int)
    helper = func (node *TreeNode) (bool, int) {
        if node == nil {
            return true, 0
        }
        
        lb, l := helper(node.Left)
        if !lb {
            return false, 0
        }
        rb, r := helper(node.Right)
        if !rb {
            return false, 0
        }
        
        b := abs(l-r) <= 1
        return b, 1 + max(l, r)
    }
    
    rb, _ := helper(root)
    
    return rb
}

func max (a, b int) int {
    if a > b {
        return a
    }
    return b
}

func abs(a int) int {
    if a < 0 {
        return -a
    }
    return a
}
